cpp/oneapi/dal/algo/knn/backend/gpu/infer_kernel_brute_force_dpc.cpp

```diff
@@ -43,27 +43,42 @@ using dal::backend::context_gpu;
 template <typename Task>
 using descriptor_t = detail::descriptor_base<Task>;
 
+using voting_t = ::oneapi::dal::knn::voting_mode;
+
 namespace de = ::oneapi::dal::detail;
 namespace bk = ::oneapi::dal::backend;
 namespace pr = ::oneapi::dal::backend::primitives;
 
-template <typename Float>
-sycl::event sqrt(sycl::queue& q, array<Float>& data, const bk::event_vector& deps = {}) {
-    ONEDAL_ASSERT(data.has_mutable_data());
-    const auto length = data.get_count();
-    const auto range = bk::make_range_1d(length);
-    auto* const data_ptr = data.get_mutable_data();
+template <typename T1, typename T2>
+sycl::event copy_with_sqrt(sycl::queue& q,
+                           pr::ndview<T1, 2>& dst,
+                           const pr::ndview<T2, 2>& src,
+                           const bk::event_vector& deps = {}) {
+    ONEDAL_ASSERT(src.has_data());
+    ONEDAL_ASSERT(dst.has_mutable_data());
+    const pr::ndshape<2> dst_shape = dst.get_shape();
+    ONEDAL_ASSERT(dst_shape == src.get_shape());
+    T1* const dst_ptr = dst.get_mutable_data();
+    const T2* const src_ptr = src.get_data();
+    const auto dst_stride = dst.get_leading_stride();
+    const auto src_stride = src.get_leading_stride();
+    const auto cp_range = bk::make_range_2d(dst_shape[0], dst_shape[1]);
     return q.submit([&](sycl::handler& h) {
         h.depends_on(deps);
-        h.parallel_for(range, [=](sycl::id<1> idx) {
-            data_ptr[idx] = sycl::sqrt(data_ptr[idx]);
+        h.parallel_for(cp_range, [=](sycl::id<2> idx) {
+            T1& dst_ref = *(dst_ptr + idx[0] * dst_stride + idx[1]);
+            const T2& val_ref = *(src_ptr + idx[0] * src_stride + idx[1]);
+            dst_ref = sycl::sqrt(val_ref);
         });
     });
 }
 
 template <typename Float>
 class knn_callback {
-    using voting_t = std::unique_ptr<pr::uniform_voting<idx_t>>;
+    using dst_t = Float;
+
+    using uniform_voting_t = std::unique_ptr<pr::uniform_voting<idx_t>>;
+    using distance_voting_t = std::unique_ptr<pr::distance_voting<dst_t>>;
```

Comment:
Why they are unique pointers?

Reply:
1) Both `uniform_voting` and `distance_voting` are abstract classes. So they should be passed with pointers.
2) These objects are not thread safe so they should have only one owner to guarantee absence of race conditions.
2) We have to use smart pointers wherever it is possible.

All facts above mean that we have to use `unique_ptr`s

Reply:
Discussed